File: src/TypeFormatter/NeverTypeFormatter.ts
import type { Definition } from "../Schema/Definition.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { NeverType } from "../Type/NeverType.js";

export class NeverTypeFormatter implements SubTypeFormatter {
    public supportsType(type: BaseType): boolean {
        return type instanceof NeverType;
    }
    public getDefinition(type: NeverType): Definition {
        return { not: {} };
    }
    public getChildren(type: NeverType): BaseType[] {
        return [];
    }
}

File: internal/formatter/never_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// NeverTypeFormatter mirrors src/TypeFormatter/NeverTypeFormatter.ts.
//
// In the TypeScript implementation HiddenType subclasses NeverType, so
// `instanceof NeverType` also matches hidden types; IsNeverLike mirrors
// that here. The Hidden formatter is registered first, so it wins for
// HiddenType instances.
type NeverTypeFormatter struct{}

func NewNeverTypeFormatter() *NeverTypeFormatter { return &NeverTypeFormatter{} }

func (f *NeverTypeFormatter) SupportsType(t types.Type) bool {
	return types.IsNeverLike(t)
}

func (f *NeverTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	return &schema.Definition{Not: &schema.Definition{}}
}

func (f *NeverTypeFormatter) GetChildren(t types.Type) []types.Type { return nil }
